lib/backend.js

```diff
@@ -568,6 +573,80 @@ Backend.prototype.getChannels = function(collection, id) {
   ];
 };
 
+Backend.prototype.getSnapshot = function(agent, index, id, version, timestamp, callback) {
+  var backend = this;
+  this._getSnapshot(agent, index, id, version, timestamp, function (error, snapshot) {
+    if (error) return callback(error);
+
+    var request = {
+      collection: index,
+      id: id,
+      version: snapshot.version,
+      timestamp: snapshot.timestamp,
+      snapshots: snapshot.data ? [snapshot.data] : [],
+      type: snapshot.type
+    };
+
+    backend.trigger(backend.MIDDLEWARE_ACTIONS.readSnapshots, agent, request, function (error) {
+      if (error) return callback(error);
+      callback(null, {
+        data: request.snapshots[0],
+        version: request.version,
+        timestamp: request.timestamp,
+        type: request.type
+      });
+    });
+  });
+};
+
+Backend.prototype._getSnapshot = function (agent, index, id, version, timestamp, callback) {
+  version = (typeof version === 'number' && isFinite(version)) ? version + 1 : null;
+  var options = { metadata: true };
+  var timestampIsNumber = typeof timestamp === 'number';
+
+  this._getOps(agent, index, id, 0, version, options, function (error, ops) {
```

Comment:
It looks like if you specify a Date rather than a version number, _all_ the ops are loaded into memory, but if you specify a version number, _only_ the ops until that version are loaded into memory, is that right?

If this is the case, it may be worth mentioning the overhead of passing in a Date in the docs.

Reply:
Yes, that's exactly right. As far as I'm aware, with the current definition for `getOps`, this is unavoidable, right? I'll update the docs.

Reply:
Yes, with the current definition for `getOps`, it appears to be unavoidable.

Although, conceivably, one could optimize that case by passing the date into the database driver, so the database query itself could specify "less than this date" (as the timestamps are there on the ops and queryable). That feels like a rabbit hole though, and would require changing the database driver API, and then implementing the change in the actual drivers. Maybe something to think about for the future, once this first iteration lands.

Thanks for updating the docs!

Reply:
In fact, it looks like all ops are always loaded into memory, see https://github.com/share/sharedb-mongo/blob/master/index.js#L362, so I'd say we don't need a specific warning about getting a snapshot for a `Date`. https://github.com/share/sharedb-mongo will need to be optimized but that would be a separate piece of work.

Reply:
Fine - will reset

Reply:
@gkubisa Oh wow, that's surprising! I would have thought sharedb-mongo would query against the version. Good catch.

Reply:
@alecgibson This fork is active https://github.com/teamwork/sharedb-mongo/ . I'm using that fork currently as my dependency, instead of the original https://github.com/share/sharedb-mongo.

I opened a new issue just now **Optimization opportunity: use "to" in getOps query** https://github.com/share/sharedb-mongo/issues/61